core/libp2p/transport/tcp/tcp_transport.cpp

```diff
@@ -58,4 +44,41 @@ namespace libp2p::transport {
   TcpTransport::TcpTransport(boost::asio::io_context &context,
                              std::shared_ptr<Upgrader> upgrader)
       : context_(context), upgrader_(std::move(upgrader)) {}
+
+  void TcpTransport::onResolved(std::shared_ptr<TcpConnection> c,
+                                const TcpConnection::ResolverResultsType &r,
+                                Transport::HandlerFunc handler) {
+    c->connect(r,
+               [self{this->shared_from_this()}, c, handler{std::move(handler)}](
+                   auto ec, auto &e) mutable {
+                 if (ec) {
+                   return handler(ec);
+                 }
+
+                 self->onConnected(std::move(c), std::move(handler));
+               });
+  }
+
+  void TcpTransport::onConnected(std::shared_ptr<TcpConnection> c,
+                                 Transport::HandlerFunc handler) {
+    this->upgrader_->upgradeToSecure(
+        std::move(c),
+        [self{shared_from_this()},
+         handler{std::move(handler)}](auto &&rsecureConn) mutable {
+          if (!rsecureConn) {
+            return handler(rsecureConn.error());
+          }
+
+          auto &&secureConn = rsecureConn.value();
+
+          return self->upgrader_->upgradeToMuxed(
+              std::forward<decltype(secureConn)>(secureConn),
```

Comment:
Maybe just
```suggestion
              std::move(rsecureConn.value()),
```
And not declaring `secureConn`?